### ISSUE TITLE
BartTokenizer for Russian language?

### ISSUE THREAD

**Issue by zzomg** [huggingface/tokenizers#889]
Hi 👋 I've fine-tuned pre-trained BART model for Russian language, and have used pre-trained BartTokenizer to encode input. Now that I'm trying to decode model predictions with the same tokenizer it gives strange output like this one:

![image](https://user-images.githubusercontent.com/55043240/150517864-ce0696c9-e7d1-4dc2-b2e6-e81341fe695e.png)

I can't really figure out where � symbols come from and why they are not being decoded. I looked up their ids and they **are** present in the tokenizer vocabulary, so they are not unknown symbols.

![image](https://user-images.githubusercontent.com/55043240/150518333-0a73e3af-4b3a-4bfd-afce-8a5e91be0bb5.png)

 Also, I've checked that train dataset doesn't contain any other symbols apart from Russian alphabet and space.

Is it correct to use pre-trained tokenizer for Russian language? It seems to encode input just fine, though, and most of the output is decoded correctly. 
Maybe something is wrong with the way I decode predictions? I tried decoding method from #268 which uses `ByteLevel()` decoder but it didn't help either.

To reproduce:

```
import os 
import glob
import copy
import json
import numpy as np
from tqdm import tqdm
from omegaconf import OmegaConf

import torch
from transformers import Trainer, TrainingArguments, AutoTokenizer, BartForConditionalGeneration, DataCollatorWithPadding, BartTokenizer

def decode_output(predicts, manifest, tokenizer):
    logits = predicts.predictions[0]
    label_ids = predicts.label_ids
    
    idx = 0
    
    for logit, label in zip(logits, label_ids):
        softmax_preds = softmax(logit)
        ids = [np.argmax(score) for score in logit]
        print(ids)
        # for _id in ids:
        #     print(f"{_id} | {tokenizer.decode([_id])}")
        dec_pred = tokenizer.decode(ids, skip_special_tokens=True)
        inpt = manifest[idx]['pred_text']
        true_lbl = manifest[idx]['text']
        print(f"INPUT: {inpt}")
        print(f"REF  : {true_lbl}")
        print(f"PRED : {dec_pred}")
        print("-----------------------------------------------------------")
        idx += 1
    
def softmax(x):
    e = np.exp(x - np.max(x))
    return e / e.sum(axis=-1).reshape([x.shape[0], 1])

class Dataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)

def load_manifest(filename):
    with open(filename) as f:
        content = f.readlines()

    manifest = [json.loads(t) for t in content]
    return manifest

def main():
    model = BartForConditionalGeneration.from_pretrained('checkpoint-236000') # trained checkpoint for facebook/bart-base
    tokenizer = BartTokenizer.from_pretrained("facebook/bart-base")

    m_path = 'file1_set.json'
    m = load_manifest(m_path)

    pred_texts, pred_labels = [], []

    for sample in m:
        pred_texts.append(sample['pred_text'])
        pred_labels.append(sample['text'])
        
    pred_text_encodings = tokenizer(pred_texts, truncation=True, padding='max_length', max_length=300)
    pred_labels_encodings = tokenizer(pred_labels, truncation=True, padding='max_length', max_length=300)['input_ids']

    pred_dataset = Dataset(pred_text_encodings, pred_labels_encodings)
    
    print("Starting traning...")
    training_args = TrainingArguments(
        output_dir='./finetuning_bart_ru_model',          # output directory
        num_train_epochs=15,              # total number of training epochs
        per_device_train_batch_size=8,  # batch size per device during training
        per_device_eval_batch_size=4,   # batch size for evaluation
        warmup_steps=1000,                # number of warmup steps for learning rate scheduler
        weight_decay=0.001,               # strength of weight decay
        logging_dir='./logs',            # directory for storing logs
        logging_steps=100,
        evaluation_strategy='steps',
        eval_steps=15000,
        gradient_accumulation_steps=2,
        do_predict=True,
        do_eval=True
        # no_cuda=True
    )
    
    trainer = Trainer(
        model=model,                         # the instantiated 🤗 Transformers model to be trained
        args=training_args,                  # training arguments, defined above
    )

    predicts = trainer.predict(test_dataset=pred_dataset) 
    decode_output(predicts, m, tokenizer)

    
if __name__ == '__main__':
    main()

```
Checkpoint: https://drive.google.com/file/d/1NqFFx-XNM3sYxEUoBXd-H3KBS3wGMI7z/view?usp=sharing
Test manifest: https://drive.google.com/file/d/1L3fsXLAZHSs2Gh6gRfO_9cxdMB9E-Cvo/view?usp=sharing
Thanks!

**Comment by Narsil** [comment]
@zzomg Do you have the correct fonts installed on your machine ?

As you can see on the screenshot of the vocab, the values already contain � . This is probably a font issue, where the font doesn't support that unicode glyph.

The first one (22063) shuold be `"Ñ"` for isntance (if you don't see N + tilde, it means it's a local issue to your computer)

**Comment by zzomg** [comment]
@Narsil I believe it's not a font issue. I saved tokenizer vocab to a .txt file and checked that everything is displayed fine, no � symbols. But then, when I save model outputs to a .txt file, �s suddenly start to appear. E.g. one of the � on the screenshot has id 11936, when I look it up in tokenizer vocabulary file it's displayed just fine:
![image](https://user-images.githubusercontent.com/55043240/150671319-31328ac5-e763-4a15-99fe-a56afb2c77c6.png)
But when it's in the model outputs file it becomes a �:
![image](https://user-images.githubusercontent.com/55043240/150671380-d02c5bb8-801a-4a0a-8e3d-f281222a709c.png)

**Comment by Narsil** [comment]
Oh, I see.

I checked a bit deeper:

```python
from transformers import AutoTokenizer
tokenizer = tokenizer = AutoTokenizer.from_pretrained("facebook/bart-base",)

tokenizer.decode([11936])
# '�'
```

So I do seem the have the same issue. However, it's not an actual issue, it's just that this token is never meant to be decoded alone.

Getting the real unicode value of `11936` it's 
```python
[k for k, v in tokenizer.get_vocab().items() if v == 11936][0]
# 'Ħ'
```
```python
tokenizer('Ħ')
# {'input_ids': [0, 649, 18164, 2], 'attention_mask': [1, 1, 1, 1]}
```

And if you try to decode `649` or `18164` on their own, you're going to have the same issue, but if you decode them togethers `tokenizer.decode([649, 18164])` then it works.

The reason is that `Bart` is a roberta/gpt2  variant, which has `ByteLevel` in the tokenizer. What that means is that real bytes all get assigned a printable code (like 649, 11836 or 18164) but while being decoded, they get attributed real byte values, which are not necessarily valid utf-8. When the resulting string is not valid, it just gets replaces with `�`.

Does that answer your question ?

**Comment by zzomg** [comment]
@Narsil Thanks for your answer! Now I see where �s come from. However, I'm still not quite sure if there is a way to get rid of them? The outputs are really hard to comprehend because of all the �.

**Comment by Narsil** [comment]
Is the model fully trained yet ?  If it's early on, it might just indicate it needs further training. The fact that is outputs bad tokens for a while is not necessarily an issue.

If it's been training for a while, it might indicate the targets have an issue (or I misunderstood the issue).

**Comment by zzomg** [comment]
@Narsil It’s been fine tuned for about 5 epochs, seems like it needs further fine tuning. 
> it might indicate the targets have an issue

What kind of issue could it be? In my case, targets are plain text.

**Comment by zzomg** [comment]
UPD: I've spent some more time debugging the model, so I trained it one a dataset that consists of one sample only and when it overfits to this one sample, � are completely gone. So the model *is* capable of getting rid of these symbols but it was not trained enough after all. Another problem with that, however, is that I'm not able to distribute training to multiple GPUs but for that I'll open another issue, I guess. Thanks for help!
